File: 199-BinaryTreeRightSideView/199-BinaryTreeRightSideView.java
// Last updated: 8/16/2026, 9:51:25 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        // append each layer's most right node -> BFS
        List<Integer> res = new ArrayList<>();
        if (root == null) return res;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while(!queue.isEmpty()) {
            // List<Integer> temp = new ArrayList<>();
            int qSize = queue.size();
            for(int i = 0; i < qSize; i ++) {
                TreeNode curr = queue.poll();
                // temp.add(curr.val);
                // the last node will be saved into "temp"
                if (i == qSize - 1) res.add(curr.val);
                
                if(curr.left != null) queue.offer(curr.left);
                if(curr.right != null) queue.offer(curr.right);
            }
            // res.add(temp.get(temp.size() - 1));
        }

        return res;
    }
}

File: 199-BinaryTreeRightSideView/199-BinaryTreeRightSideView.py
# Last updated: 9/16/2026, 12:13:55 PM
1# Definition for a binary tree node.
2# class TreeNode:
3#     def __init__(self, val=0, left=None, right=None):
4#         self.val = val
5#         self.left = left
6#         self.right = right
7class Solution:
8    def rightSideView(self, root: TreeNode | None) -> list[int]:
9        # level order BFS, pick the last element
10        if root is None:
11            return []
12
13        q = deque([root])
14        res = []
15        while q:
16            q_size = len(q)
17            for i in range(q_size):
18                node = q.popleft()
19                if i == q_size - 1:
20                    # last item
21                    res.append(node.val)
22                
23                if node.left: q.append(node.left)
24                if node.right: q.append(node.right)
25        
26        return res
27        